api/resolvers/index.js

```diff
@@ -1651,15 +1651,116 @@ const resolvers = {
+        ] = await Allocation.aggregate([
+          {
+            $match: {
+              eventMemberId: mongoose.Types.ObjectId(eventMemberId),
+            },
+          },
+          { $group: { _id: null, totalAllocations: { $sum: "$amount" } } },
+        ]);
+
+        const [
+          { totalContributions } = { totalContributions: 0 },
+        ] = await Contribution.aggregate([
+          {
+            $match: {
+              eventMemberId: mongoose.Types.ObjectId(eventMemberId),
+            },
+          },
+          { $group: { _id: null, totalContributions: { $sum: "$amount" } } },
+        ]);
+
+        const balance = totalAllocations - totalContributions;
+
+        const adjustment = amount - balance;
+
+        await new Allocation({
+          organizationId: currentOrgMember.organizationId,
+          eventId: targetEventMember.eventId,
+          eventMemberId,
+          amount: adjustment,
+        }).save();
+      }
 
       return targetEventMember;
     },
+    bulkAllocate: async (
+      _,
+      { eventId, amount, type },
+      { currentOrgMember, models: { EventMember, Allocation, Contribution } }
+    ) => {
+      if (!currentOrgMember) throw new Error("You need to be logged in.");
+
+      const eventMembers = await EventMember.find({
+        eventId,
+        isApproved: true,
+      });
+
+      const currentEventMember = await EventMember.findOne({
+        orgMemberId: currentOrgMember.id,
+        eventId,
+      });
+
+      if (!currentEventMember?.isAdmin)
+        throw new Error("You need to be event admin to allocate funds.");
+
+      for (const member of eventMembers) {
+        if (type === "ADD") {
+          await new Allocation({
+            organizationId: currentOrgMember.organizationId,
+            eventId,
+            eventMemberId: member.id,
+            amount,
+          }).save();
+        } else if (type === "SET") {
+          const [
+            { totalAllocations } = { totalAllocations: 0 },
+          ] = await Allocation.aggregate([
+            {
+              $match: {
+                eventMemberId: mongoose.Types.ObjectId(member.id),
+              },
+            },
+            { $group: { _id: null, totalAllocations: { $sum: "$amount" } } },
+          ]);
+
+          const [
+            { totalContributions } = { totalContributions: 0 },
+          ] = await Contribution.aggregate([
+            {
+              $match: {
+                eventMemberId: mongoose.Types.ObjectId(member.id),
+              },
+            },
+            { $group: { _id: null, totalContributions: { $sum: "$amount" } } },
+          ]);
+
+          const balance = totalAllocations - totalContributions;
+
+          const adjustment = amount - balance;
+
+          await new Allocation({
+            organizationId: currentOrgMember.organizationId,
+            eventId,
+            eventMemberId: member.id,
+            amount: adjustment,
+          }).save();
+        }
```

Comment:
This chunk of code inside the for loop looks like it's mostly the same as the code in the other `allocate` resolver. Would be nice to extract the common logic into a shared function.

Reply:
I consolidated into a function that I put i `api/controller/index.js` where we can add more controller type functionality, inspired by #75

Reply:
Beautiful!